doc/Userland.md

```diff
@@ -80,41 +99,66 @@ command(uint32_t driver, uint32_t command, int data) {
-passed as the pointer `userdata`. After the application has started the timer,
-calls `yield`, and the timer fires, the callback function will be called.
-
-It is important to note that `yield` must be called for events to be serviced in
-the current implementation of Tock. Callbacks to the application will be queued
-when they occur but the application will not receive them until it yields. This
-is not fundamental to Tock, and future version may service callbacks on any
-system call or when performing application time slicing. After receiving and
-running the callback, application code will continue after the `yield`.
-Applications which are "finished" (i.e. have returned from `main()`) should call
-`yield` in a loop to avoid being scheduled by the kernel.
-
+A detailed description of Tock's system call API and ABI can be found in
+[TRD104](reference/trd-syscalls.md). The [system call
+documentation](./Syscalls.md) describes how the are implemented in the
+kernel.
+
+
+## Upcalls and Termination
+
+The Tock kernel is completely non-blocking, and it pushes this 
+asynchronous behavior to userspace code. This means that system calls
+(with one exception) do not block. Instead, they always return very quickly.
+Long-running operations (e.g., sending data over a bus, samping a sensor)
+signal their completion to userspace through upcalls. An upcall is a function 
+call the kernel makes on userspace code.
+
+Yield system calls are the exception to this non-blocking rule. The yield-wait 
+system call blocks until the kernel invokes an upcall on the process. 
+The kernel only invokes upcalls when a process issues the yield system call:
+it does not invoke upcalls at arbitrary points in the program.
+
+For example, consider the case of when a process wants to sleep for 100 
+milliseconds. The timer library might break this into three operations:
+
+1. It registers an upcall for the timer system call driver with a Subscribe
+system call.
+2. It tells the timer system call driver to issue an upcall in 100 
+milliseconds by invoking a Command system call.
+3. It calls the yield-wait system call. This causes the process to block
+until the timer upcall executes. The kernel pushes a stack frame onto
+the process to execute the upcall; this function call returns to the
+instruction after yield was invoked.
+
+When a process registers an upcall with a call to a Subcribe system call,
+it may pass a pointer `userdata`. The kernel does not access or use this
+data: it simply passes it back on each invocation of the upcall. This
+allows a process to register the same function as multiple upcalls, and
+distinguish them by the data passed in the argument.
+
+It is important to note that upcalls are not executed until a process
+calls `yield`. The kernel will enqueue upcalls as events occur within
+the kernel, but the application will not handle them until it yields.
+
+Applications which are "finished" (i.e. have returned from `main()`)
```

Comment:
```suggestion
Applications which are "finished"
```

Don't think we want to use 'return from `main`' as example, since the default userspace behavior is to spin in a loop waiting for events rather than to call Exit on the application's behalf -- this example confuses that I think.